### ISSUE TITLE
Syntax Feature Request: Skippable Groups

### ISSUE THREAD

**Issue by amadeus** [vim/vim#872]
I have come across a specific syntax regex feature that would be very useful. I help maintain a very popular javascript syntax plugin and this feature request has come out of this work.

As it stands, there's a very helpful feature for assigning `nextgroup`s when you add the `skipempty` and `skipwhite` options. I think it would be very helpful to add another option `skipgroups`, that would then apply to groups that have been marked as skippable.

An example:

``` js
if (someVar > 2) {
}
```

In this example, we can do some fancy stuff where we say: `nextgroup` on `if` is an `ifElseParens` region. Then from there, we can do a `nextgroup` on `ifElseParens` for an `ifElseRegion` defined by the `{` and `}`.

`skipwhite` and `skipempty` of course let us handle the multitude of various spacing cases that could occur, for example:

``` js
if(someVar > 2){}

if
(someVar > 2)
{
}

if (someVar > 2)
{
}

if(someVar > 2)
    {
    }
```

These are all perfectly valid javascript and would be perfectly matched using `skipwhite`, `skipempty` and `nextgroup`.

However, things get much more hairy if we throw a comment in there:

``` js
if(someVar > 2) // This is a comment
{
}
```

Now we have a case where `IfElseRegion` won't be matched using `nextgroup` because there's a comment there.

This is still valid JS however. As it stands, the only real fix is to create new special comments that are contained and ONLY matched as `nextgroup`s to then continue the region matching afterwards. It means we'd have to create a good number of them to fix various specially designed regions, because the comments are valid.

I think a far more elegant fix, would be something like this:

```
syntax region  jsComment start=/\/\*/ end=/\*\// skippable
syntax keyword jsIfStatement if skipwhite skipempty skipgroups nextgroup=jsIfElseParens
syntax region  jsIfElseParens start=/(/ end=/)/ skipwhite skipempty skipgroups nextgroup=jsIfElseRegion
```

Basically `skippable` would pool that `jsComment` region to a special classification of groups that could then be opted into, as something to be skipped when testing `nextgroup`, and it would only apply if `skipgroups` was added as a possible skip option.

I think this could really help many syntax maintainers more easily create more accurate syntax matching that is also easier to maintain.

**Comment by amadeus** [comment]
Any thoughts on this? I feel like it could be a hugely invaluable feature to making syntax highlighting far more precise and accurate.

Another possible API could be:

```
syntax region  jsIfElseParens start=/(/ end=/)/ skipwhite skipempty skipgroup=jsComment nextgroup=jsIfElseRegion
```

A `skipgroup` that functions like `nextgroup`, but for what would be skippable.

**Comment by bb010g** [comment]
I just ran into a similar desire for `skipgroup` and `endgroup`, somewhat overlapping with #1714. Thoughts on semantics:

## `:syn-region` arguments `:syn-endgroup` & `:syn-skipgroup`

```vimhelp
DEFINING REGIONS	*:syn-region* *:syn-start* *:syn-skip* *:syn-end*
							*E398* *E399*
:sy[ntax] region {group-name} [{options}]
		[matchgroup={group-name}]
		[keepend]
		[extend]
		[excludenl]
		start={start_pattern} ..
		[skip={skip_pattern}]
		[skipgroup={group-name}]
		end={end_pattern} ..
		[endgroup={group-name}]
		[{options}]
```

(I'm not sure how either-or-both should be signified in the manual for `end=` & `endgroup=`; normally the summary would be duplicated but this one is rather long.)

```vimhelp
	skip={skip_pattern}	The search pattern that defines text inside
				the region where not to look for the end
				pattern.  See |:syn-pattern| below.
	skipgroup={group-name}	The syntax group to match inside the item and
				don't allow the end pattern to begin inside.
				Used where the skip pattern is. See
				|:syn-skipgroup|.
	end={end_pattern}	The search pattern that defines the end of
				the region.  See |:syn-pattern| below.
	endgroup={group-name}	The syntax group to match inside the item and
				end the region when it ends.  Used where the
				end pattern is. See |:syn-endgroup|.
```

The next two sections follow `:syn-keepend`.

```
							*:syn-endgroup*
	endgroup={group-name},...

	The "endgroup" argument is followed by a list of syntax group names.
	By default, a contained match can obscure a match for the end pattern.
	If you don't want this, the "endgroup" argument will make the matching
	of any of its named syntax groups end their containing region when
	they end, as long as the match begins after the start pattern.
	This works as if this region has a "contains=" argument that includes
	these names.

	The {group-name}... can be used just like for "contains", as explained
	above.

	To contrast "keepend" and "endgroup":
	- A region with "keepend" ends with an end pattern match's end,
	  so contained items may not extend the region and skip or end
	  patterns may not match again.
	- A region with "endgroup" ends with an end syntax group match's end,
	  so skip or end patterns may not extend the region and contained
	  items may not match again.


							*:syn-skipgroup*
	skipgroup={group-name},...

	The "skipgroup" argument is followed by a list of syntax group names.
	By default, a contained match can obscure a match for the end pattern.
	However, the "keepend" argument will make the matching of the end
	pattern also end any contained item.
	If you want more control over this, the "skipgroup" argument releases
	these listed groups from this restriction, so only a contained match
	from a group mentioned in "skipgroup" argument can obscure a match for
	the end pattern.
	This works as if this region has a "contains=" argument that includes
	these names.

	The {group-name}... can be used just like for "contains", as explained
	above.
```

I don't believe these arguments run counter to Vim's syntax highlighting philosophy. Contained groups are mostly ran the same. `"keepend"` interactions will differ. The early ending of the group will pull back an existing `"end="` match, never search fresh past its end.

`"endgroup="` does not exempt its items from `keepend`; instead, those items may also be mentioned in `"skipgroup="`.

Here's an example usage from a Forth-type language, Factor. We want to match code of the form `GENERIC#: <word> <number>` in a region, and allow for any syntax group in `@factorComment` to match in-between. The members of syntax cluster `@factorComment` is declared with properly higher priority than other Factor syntax groups. `syn iskeyword` has been initialized with the valid (non-whitespace) token characters. The syntax group `factorIntegerRequired` matches a token that isn't an integer and is linked to the `Error` highlight group.

```vim
syn  region factorGenericN matchgroup=factorGenericN start=/\v<GENERIC\#:>/ matchgroup=NONE end=/\v\S/ endgroup=factorGenericNContent contains=@factorComment
syn  region factorGenericNContent matchgroup=factorGenericN start=/\v<\S+>/ matchgroup=NONE end=/\v<\S+>/ contains=@factorComment,factorIntegerRequired
```

As far as I know, equivalent functionality to this, where any contained `@factorComment` match (an arbitrary set of syntax groups) extends these regions, and the initial region still terminates, is impossible in current Vim. Ideally, `factorGenericNContent` would also be transparent, but then the start group would match `factorIntegerRequired`, and I don't know of a way to reinstate `matchgroup`'s restriction. You could possibly get around this with some more syntax groups of more perverse code.

To contrast, a `GENERIC: <word>` region *is* possible currently, with:

```vim
syn  region factorGeneric matchgroup=factorGeneric start=/\v<GENERIC:>/ matchgroup=NONE end=/\v<\S+>/ contains=@factorComment
```

The multiple matching of arbitrary syntax cluster sets inside the region is where you run into problems. Matching an arbitrary syntax cluster once is no problem.

**Comment by axvr** [comment]
We are also in need of something like this for clojure-vim/clojure.vim#17.
